jumanji/environments/packing/flat_pack/generator.py

```diff
@@ -0,0 +1,402 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Tuple
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.packing.flat_pack.types import State
+from jumanji.environments.packing.flat_pack.utils import (
+    compute_grid_dim,
+    get_significant_idxs,
+    rotate_block,
+)
+
+
+class InstanceGenerator(abc.ABC):
+    """Base class for generators for the flat_pack environment. An `InstanceGenerator` is responsible
+    for generating a problem instance when the environment is reset.
+    """
+
+    def __init__(
+        self,
+        num_row_blocks: int,
+        num_col_blocks: int,
+    ) -> None:
+        """Initialises a flat_pack generator, used to generate grids for the FlatPack environment.
+
+        Args:
+            num_row_blocks: Number of row blocks in flat_pack environment.
+            num_col_blocks: Number of column blocks in flat_pack environment.
+        """
+
+        self.num_row_blocks = num_row_blocks
+        self.num_col_blocks = num_col_blocks
+
+    @abc.abstractmethod
+    def __call__(self, key: chex.PRNGKey) -> State:
+        """Call method responsible for generating a new state.
+
+        Args:
+            key: jax random key in case stochasticity is used in the instance generation process.
+
+        Returns:
+            A `FlatPack` environment state.
+        """
+
+
+class RandomFlatPackGenerator(InstanceGenerator):
+    """Random flat_pack generator. This generator will generate a random flat_pack grid."""
+
+    def __init__(
+        self,
+        num_row_blocks: int,
+        num_col_blocks: int,
+    ):
+        """Initialises a random flat_pack generator.
+        Args:
+
+            num_row_blocks: Number of row blocks in flat_pack grid.
+            num_col_blocks: Number of column blocks in flat_pack grid.
+        """
+        super().__init__(num_row_blocks, num_col_blocks)
```

Comment:
Same as the default so you can just leave this out
```suggestion
```

Reply:
Done